decompiler/pipeline/ssa/dependency_graph.py

```diff
@@ -1,74 +1,64 @@
-        return dependency_graph
+def _collect_variables(cfg: ControlFlowGraph) -> Iterator[Variable]:
+    for instruction in cfg.instructions:
+        for subexpression in instruction.subexpressions():
+            if isinstance(subexpression, Variable):
+                yield subexpression
 
-    def _non_interfering_requirements(self, requirements: List[Variable], defined_variable: Variable) -> Optional[Variable]:
-        """Get the unique non-interfering requirement if it exists, otherwise we return None."""
-        non_interfering_requirement = None
-        for required_variable in requirements:
-            if self._variables_can_have_same_name(defined_variable, required_variable):
-                if non_interfering_requirement:
-                    return None
-                non_interfering_requirement = required_variable
-        return non_interfering_requirement
 
-    def _variables_can_have_same_name(self, source: Variable, sink: Variable) -> bool:
-        """
-        Two variable can have the same name, if they have the same type, are both aliased or both non-aliased variables, and if they
-        do not interfere.
+def _assignments_in_cfg(cfg: ControlFlowGraph) -> Iterator[Assignment]:
+    """Yield all interesting assignments for the dependency graph."""
+    for instr in cfg.instructions:
+        if isinstance(instr, Assignment):
+            yield instr
+
 
-        :param source: The potential source vertex.
-        :param sink: The potential sink vertex
-        :return: True, if the given variables can have the same name, and false otherwise.
-        """
-        if self.interference_graph.are_interfering(source, sink) or source.type != sink.type or source.is_aliased != sink.is_aliased:
-            return False
-        if source.is_aliased and sink.is_aliased and source.name != sink.name:
-            return False
-        return True
+def _expression_dependencies(expression: Expression) -> dict[Variable, float]:
+    match expression:
+        case Variable():
+            return {expression: 1.0}
+        case Operation():
+            operation_type_penalty = {
+                OperationType.call: 0,
+                OperationType.address: 0,
+                OperationType.dereference: 0,
+                OperationType.member_access: 0,
+            }.get(expression.operation, 0.5)
 
-    def get_components(self) -> Iterable[Set[Variable]]:
-        """Returns the weakly connected components of the dependency graph."""
-        for component in weakly_connected_components(self):
-            yield set(component)
+            operands_dependencies = (_expression_dependencies(operand) for operand in expression.operands)
+            dependencies: dict[Variable, float] = reduce(dict.__or__, operands_dependencies, {})
+            for var in dependencies:
+                dependencies[var] /= len(dependencies)
+                dependencies[var] *= operation_type_penalty
```

Comment:
I am not sure about these computations. 
A few examples:
1. `x = a + 2` --> `operation_type_penalty = 0.5` and `operand_dependencies = ({a : 1}, {})` --> `dependencies = {a: 0.5}`
2. `x = a + b` --> `operation_type_penalty = 0.5` and `operand_dependencies = ({a : 1}, { b:1})` --> `dependencies = {a: 0.25, b: 0.25}`
3. 2. `x = a + 2*a` --> `operation_type_penalty = 0.5` and `operand_dependencies = ({a: 1}, { a: 0.5})` --> `dependencies = {a: 0.25}` --> This should have value 0.5 or?
4. `x = a + a*b` --> `operation_type_penalty = 0.5` and `operand_dependencies = ({a : 1}, {a:0.5, b:0.5})` --> `dependencies = {a: 0.25, b: 0.25}` --> Do you want that a and b have the same value here?

Second, why not immediately return the empty dictionary, if `operation_type_penalty` is 0? Because then all values are zero?

Reply:
I changed the algorithm with f02b7a7b.